### ISSUE TITLE
Unique results for unique places

### ISSUE THREAD

**Issue by AvidSeeker** [organicmaps/organicmaps#4165]
[Relevant Reddit post](https://www.reddit.com/r/openstreetmap/comments/zwvyb8/unique_results_for_unique_places/)

For some reason, OM has different result order from OSM.

It's annoying to search for some unique place, and have a bunch of irrelevant results. For example, try searching for "Eiffel tower" in OM. You still have to manually search in the results to find what your meaning is.

Would be convenient to have OM automatically selects the most relevant result, and if for some reason the user isn't satisfied by the default top result, then he can fine-tune his search.

**Comment by biodranik** [comment]
Let's focus on OM here. OSM uses nominatim and other address databases for search. We only use our own search algo on top of the OSM data.

To make search results relevant, we need to tune the algorithm. To do that properly, we need your use cases/examples:
1. Which maps/regions were downloaded?
2. What did you type?
3. What was your current position at that moment?
4. What was on the map screen at that moment?
5. What did you see in the search results list?
6. What did you expect to see in the search results list?

With that information, the algorithm can be tuned, fixed, and tested.

**Comment by AvidSeeker** [comment]
Here's an example.

1. Downloaded region: USA
2. Search: "UC Berkeley"
3. Location disabled
4. Screen location doesn't matter. Results are the same.
5. OM results:
```
Berkeley Avenue (Irvine)
Berkeley Way (Yorba Linda)
Berkeley Street (Montclair)
Berkeley Avenue
```
6. OSM results (two results):
```
University [University of California, Berkeley, East West Trail, Oakland, Alameda County, CAL Fire Northern Region, California, 94720, United States]

Bus Stop [UC Berkeley, The Crescent, Downtown Berkeley, Berkeley, Alameda County, CAL Fire Northern Region, California, 94704, United States]
```
7. Expected result: since "UC Berkeley" refers to a unique place, OM should focus directly by default on that unique place. If user was referring to a street or another building, he can fine-tune his search.

Same goes for other unique places like "Eiffel tower", "Statue of liberty", "Harvard university", etc.

**Comment by biodranik** [comment]
Tag `short_name=UC Berkeley` is not indexed yet by Organic Maps. The issue will be fixed when it will be indexed. What happens if you type `University of California, Berkeley` ?

```
 // In future we can use this tags for improve our search
  SKIP_KEY_BY_PREFIX("nat_name");
  SKIP_KEY_BY_PREFIX("reg_name");
  SKIP_KEY_BY_PREFIX("loc_name");
  SKIP_KEY_BY_PREFIX("lock_name");
  SKIP_KEY_BY_PREFIX("local_name");
  SKIP_KEY_BY_PREFIX("short_name");
  SKIP_KEY_BY_PREFIX("official_name");
```

**Comment by AvidSeeker** [comment]
University of California, Berkeley give the following (in that order):

```
Berkeley Avenue
University of California, Irvine
Office of Advancement, University of California, Irvine
...
```

totally irrelevant results. Looks like OM is confused when a place name is lengthy. Here is another example:

```
// University of California, San Diego
San Diego Freeway
San Diego Creek Trail
San Diego Creek Trail - Regional Parks Connector Bikeway
```

Can you reproduce?

**Comment by AvidSeeker** [comment]
@biodranik 
> OSM uses nominatim and other address databases for search

What are the other address databases?

Is there a way to contribute to enhance search results in these databases until OM implements them? I still have other search entries that are badly indexed.

**Comment by biodranik** [comment]
Check Nominatim docs: https://nominatim.org/admin/Import/

You can improve OM's search by adding missing addresses in OpenStreetMap.org

**Comment by AvidSeeker** [comment]
What are the results showing for "University of California, Berkeley" and "University of California, San Diego"? I get these results:

> University of California, Berkeley give the following (in that order):
> 
> ```
> Berkeley Avenue
> University of California, Irvine
> Office of Advancement, University of California, Irvine
> ...
> ```
> 
> totally irrelevant results. Looks like OM is confused when a place name is lengthy. Here is another example:
> 
> ```
> // University of California, San Diego
> San Diego Freeway
> San Diego Creek Trail
> San Diego Creek Trail - Regional Parks Connector Bikeway
> ```
> 
> Can you reproduce?

**Comment by biodranik** [comment]
OM's results are directly influenced by:
1. Your current position (if it's known)
2. The piece of map that is visible on the screen.
3. The presence of the town/city in the search query where the object is located.

So you will get different results depending on these factors.

**Comment by vng** [comment]
@AvidSeeker Ok than please wait for the next update. Looks like it's fixed on my working version.

**Comment by AvidSeeker** [comment]
> OM's results are directly influenced by:
>  ...
> So you will get different results depending on these factors.

Yes, exactly. And my issue is that there are unique places that shouldn't be affected by these factors.

**Comment by AvidSeeker** [comment]
> Ok than please wait for the next update.

Thank you. I'll try to give more test cases if found.

**Comment by biodranik** [comment]
Any new test cases?

**Comment by avidseeker** [comment]
As of `2023-12.20-4-FDroid` version, OM still shows irrelevant results:

```
UC Berkeley
---
Berkeley Street
Berkeley Street
Berkeley Drive
Berkeley Court
...
```
and
```
Eiffel tower
---
Eiffel Tower Cafe
Eiffel Tower
Eiffel Circle
...
```
